keras_cv/losses/box_3d_regression_loss.py

```diff
@@ -0,0 +1,129 @@
+
+
+def l1(y_true, y_pred, sigma=9.0):
+    """Computes element-wise l1 loss."""
+
+    absolute_difference = tf.abs(y_pred - y_true)
+    loss = tf.where(
+        absolute_difference < 1.0 / sigma,
+        0.5 * sigma * absolute_difference**2,
+        absolute_difference - 0.5 / sigma,
+    )
+
+    return tf.reduce_sum(loss, axis=-1)
+
+
+class Box3DRegressionLoss(keras.losses.Loss):
+    """Implements a bin-based box regression loss for 3D bounding boxes.
+
+    Reference: https://arxiv.org/abs/1812.04244
+
+    Box3DRegressionLoss uses L1 loss on the individual components of boxes, with
+    the exception of the bin-based heading component of each box, where the bin
+    indicator outputs use cross entropy loss, and the heading residual uses L1
+    loss.
+
+    Ground truth boxes are expected to follow the CENTER_XYZ_DXDYDZ_PHI format.
+    Refer to https://github.com/keras-team/keras-cv/blob/master/keras_cv/bounding_box_3d/formats.py
+    for more details on supported bounding box formats.
+
+    Box predictions are expected to be in CenterPillar heatmap-encoded format.
+
+    Args:
+        num_heading_bins: int, number of bins used for predicting box heading.
+        anchor_size: list of 3 ints, anchor sizes for the x, y, and z axes.
+    """
+
+    def __init__(self, num_heading_bins, anchor_size, **kwargs):
+        super().__init__(**kwargs)
+        self.num_heading_bins = num_heading_bins
+        self.anchor_size = anchor_size
+
+    def heading_regression_loss(self, heading_true, heading_pred):
+        # Set the heading to within 0 -> 2pi
+        heading_true = tf.math.floormod(heading_true, 2 * math.pi)
+
+        # Divide 2pi into bins. shifted by 0.5 * angle_per_class.
+        angle_per_class = (2 * math.pi) / self.num_heading_bins
+        shift_angle = tf.math.floormod(
+            heading_true + angle_per_class / 2, 2 * math.pi
+        )
+
+        heading_bin_label_float = tf.math.floordiv(shift_angle, angle_per_class)
+        heading_bin_label = tf.cast(heading_bin_label_float, dtype=tf.int32)
+        heading_res_label = shift_angle - (
+            heading_bin_label_float * angle_per_class + angle_per_class / 2.0
+        )
+        heading_res_norm_label = heading_res_label / (angle_per_class / 2.0)
+
+        heading_bin_one_hot = tf.one_hot(
+            heading_bin_label, self.num_heading_bins
+        )
+        loss_heading_bin = tf.nn.softmax_cross_entropy_with_logits(
+            labels=heading_bin_one_hot,
+            logits=heading_pred[..., : self.num_heading_bins],
+        )
+        loss_heading_res = l1(
+            tf.reduce_sum(
+                heading_pred[..., self.num_heading_bins :]
+                * heading_bin_one_hot,
+                axis=-1,
+                keepdims=True,
+            ),
+            heading_res_norm_label[..., tf.newaxis],
+        )
+
+        return loss_heading_bin + loss_heading_res
+
+    def regression_loss(self, y_true, y_pred):
```

Comment:
I feel like this class is doing 2~ things at once:

- unpackaging formats
- computing loss on these formats.

I feel something analogous to this is in the RetinaNet's compile method.  Instead of having a generic `BoxRegressionLoss`, we take 2 losses to compile(), `box_loss`, `classification_loss`.  Is the reason that this is packaged together due to inter-dependencies?

Perhaps it can be refactored to something like:

```
Box3DLoss(heading_loss='...', box_loss='...',)
```

Or we can attach some of the logic to the CenterPillar model instead.

Thoughts?  Am I missing some context as to why this wouldn't work, or why it is optimal to keep it all in one place?

Reply:
This is a bit of a complex issue -- I agree we should think about it.

CenterNet will already take two losses:
- Heatmap loss
- Box regression loss

IMO it makes sense for this to be one loss, because fundamentally the whole thing represents a regression loss for 3D boxes. The reason that the headings are handled separately is simply that the headings are predicted using bins instead of a single regression target, which makes the code more complex.

Happy to chat about this more. There's also an additional concept that is sometimes added to the box loss (which we'll want to support) which is a 3D IoU component. In that case, it may make sense to include this as a separate loss function.